File: wagerproof-ios-native/Wagerproof/Features/Agents/Components/AgentCardGlyphTexture.swift
import SwiftUI
import WagerproofDesign

/// Subtle animated pixel-glyph wash for the agent list cards — a scaled-down
/// echo of the `PixelWaveBackground` field on the auth gate and the agent
/// detail hero. The glyphs bloom and poof in the agent's brand hue across the
/// whole card (a touch stronger on the avatar side), giving every row a "sneak
/// peek" of the detail-page pixelwave that ties the agent UI together.
///
/// Reuses `PixelGlyphField` verbatim (same automaton, same 300ms beat) and just
/// re-tints, re-seeds, and masks it for the card. Each instance passes a stable
/// per-agent `seed`, so no two cards bloom in lockstep. Drop it into a card's
/// `.background` ZStack above the material fill; it's inert
/// (`allowsHitTesting(false)`) so the row's tap/long-press still win. See
/// `AgentDetailHero` for the full-screen version and `PixelGlyphField` for the
/// automaton.
struct AgentCardGlyphTexture: View {
    /// Raw `avatar_color` (hex or "gradient:…") — the hue the glyphs glow in.
    let avatarColor: String
    /// Stable per-card string (the agent id) → distinct glyph pattern per card.
    var seedString: String = ""
    var cornerRadius: CGFloat = 26

    @Environment(\.colorScheme) private var colorScheme

    private var primary: Color { AgentColorPalette.primary(for: avatarColor) }

    /// FNV-1a hash of the seed string → a non-zero PRNG seed. Mirrors the form
    /// chart's hashing (`String.hashValue` is per-process randomized, so it
    /// can't be used for a stable-across-launch pattern).
    private var seedValue: UInt64 {
        guard !seedString.isEmpty else { return 0x5EED_1234 }
        var h: UInt64 = 0xcbf2_9ce4_8422_2325
        for byte in seedString.utf8 {
            h = (h ^ UInt64(byte)) &* 0x0000_0100_0000_01b3
        }
        return h | 1
    }

    var body: some View {
        // Dark mode mirrors the hero exactly: white glyphs that mix toward the
        // agent hue at their cores, glowing on the dark glass. Light mode can't
        // show white, so the glyphs ARE the agent hue and deepen at their cores.
        let base: Color = colorScheme == .dark ? .white : primary
        let accent: Color = colorScheme == .dark ? primary : primary.shaded(by: 0.72)

        PixelGlyphField(
            intervals: [0.3],   // same 300ms beat as the hero field
            baseColor: base,
            accentColor: accent,
            spacing: 22,        // a touch tighter than the hero (26) for the smaller card
            dotSize: 5,
            peakOpacity: colorScheme == .dark ? 0.4 : 0.46,
            seed: seedValue
        )
        .mask(spreadMask)
        // Belt-and-suspenders clip; the host card also clips to its own shape.
        .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
        .allowsHitTesting(false)
    }

    /// Spans the full card so glyphs pulse across the entire container, but eases
    /// down toward the trailing edge — the avatar's colored side reads as the
    /// origin, and the form chart / record on the right stay legible. Never
    /// reaches zero, so the texture is present edge to edge.
    private var spreadMask: some View {
        LinearGradient(
            stops: [
                .init(color: .white, location: 0.0),
                .init(color: .white.opacity(0.84), location: 0.5),
                .init(color: .white.opacity(0.64), location: 1.0)
            ],
            startPoint: .leading,
            endPoint: .trailing
        )
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/components/AgentCardGlyphTexture.kt
package com.wagerproof.app.features.agents.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.CompositingStrategy
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.wagerproof.app.features.agents.AgentColorPalette
import com.wagerproof.core.design.backgrounds.PixelGlyphField

/**
 * Subtle animated pixel-glyph wash for agent list cards — port of iOS
 * `AgentCardGlyphTexture.swift`. Reuses the shared [PixelGlyphField] automaton
 * verbatim, re-tinted/re-seeded per agent so no two cards bloom in lockstep,
 * then masked to ease down toward the trailing edge and clipped to the card.
 *
 * Dark-only (the Android app): white glyphs mixing toward the agent hue at
 * their cores, matching the detail-hero field.
 */
@Composable
fun AgentCardGlyphTexture(
    avatarColor: String,
    seedString: String,
    modifier: Modifier = Modifier,
    cornerRadius: Dp = 26.dp,
) {
    val primary = AgentColorPalette.primary(avatarColor)
    val seed = fnv1aSeed(seedString)

    Box(
        modifier
            .clip(RoundedCornerShape(cornerRadius))
            // Offscreen so the leading→trailing fade masks the whole field's
            // alpha, not each glyph's draw call.
            .graphicsLayer(compositingStrategy = CompositingStrategy.Offscreen)
            .drawWithContent {
                drawContent()
                drawRect(
                    brush = Brush.horizontalGradient(
                        0f to Color.White,
                        0.5f to Color.White.copy(alpha = 0.84f),
                        1f to Color.White.copy(alpha = 0.64f),
                    ),
                    blendMode = BlendMode.DstIn,
                )
            },
    ) {
        PixelGlyphField(
            modifier = Modifier.fillMaxSize(),
            intervals = listOf(0.3), // same 300ms beat as the hero field
            baseColor = Color.White,
            accentColor = primary,
            spacing = 22f, // a touch tighter than the hero (26) for the smaller card
            dotSize = 5f,
            peakOpacity = 0.4,
            seed = seed,
            tapRipples = false,
        )
    }
}

/**
 * FNV-1a hash of the seed string → non-zero PRNG seed. Mirrors the iOS texture's
 * hashing; empty seed falls back to the shared hero constant.
 */
private fun fnv1aSeed(seedString: String): Long {
    if (seedString.isEmpty()) return 0x5EED_1234L
    var h = 0xcbf29ce484222325uL
    for (b in seedString.encodeToByteArray()) {
        h = (h xor b.toUByte().toULong()) * 0x100000001b3uL
    }
    return (h or 1uL).toLong()
}
